src/backend/online_monitor.ts

```diff
@@ -0,0 +1,129 @@
+import { ConnectivityStatus } from '../common/types'
+import { BrowserWindow, ipcMain, IpcMainEvent } from 'electron'
+import { logInfo, LogPrefix } from './logger/logger'
+import axios from 'axios'
+
+let status: ConnectivityStatus
+let abortController: AbortController
+let retryTimer: NodeJS.Timeout
+let retryIn = 0
+const defaultTimeBetweenRetries = 5
+let timeBetweenRetries = defaultTimeBetweenRetries
+
+// handle setting the status, dispatch events for backend and frontend, and trigger pings
+const setStatus = (newStatus: ConnectivityStatus) => {
+  logInfo(`Connectivity: ${newStatus}`, { prefix: LogPrefix.Connection })
+
+  status = newStatus
+
+  // events
+  const mainWindow = BrowserWindow.getAllWindows()[0]
+  if (mainWindow) {
+    mainWindow.webContents.send('connectivity-changed', { status, retryIn })
+  }
+  ipcMain.emit(status)
```

Comment:
I'm not sure we should use `ipcMain.emit` to trigger the `ipcMain.once('online', () => callback())` handler on line 120. The purpose of ipcMain is communication with the renderer process so I think this is confusing and forces us to mix in handlers for events passed between backend code with the ones passed between the main and renderer processes. It also calls ipcMain.emit for the other ConnectivityStatus states which would be confusing if someone declared a ipcMain.on handler with the same event name thinking that, since this wasn't used in a preload script, it was free.

This pattern also exposes excess functionality to the `ipcMain.on` handler to the preload script which hurts security.

What if we used nodejs event emitters https://nodejs.org/api/events.html for these backend only events? Then we can leave `ipcMain.on` and `ipcMain.handle` handlers solely for renderer -> main communication.

Reply:
I was using ipcMain since it is a node Event Emitter itself and it's already available everywhere. what would be the alternative? to have some file exporting another event emitter and importing it in the places we want to emit/listen events within the backend? I don't know best practices for these messages (I literally just learned about this for this PR xD)

Reply:
if you want to add that to this PR go ahead!

Reply:
> I was using ipcMain since it is a node Event Emitter itself and it's already available everywhere. what would be the alternative? to have some file exporting another event emitter and importing it in the places we want to emit/listen events within the backend? I don't know best practices for these messages (I literally just learned about this for this PR xD)

Do you think that we will want to listen to these status events in other files or will it be contained only to `online_monitor.ts`? If it's contained to this file, the event emitter can just be local to the file. 

If not, then perhaps we could export the event emitter from this module and import it where needed? Could create dependency issues though so maybe we could have a separate module that only holds the event emitters.

Reply:
I think it's useful to be able to listen to this event anywhere that's needed (like maybe the downloads manager can listen to this status changes to restart a download when the connection is back)

Reply:
I added an exported `connectivityEmitter` from the monitor file. If it creates any issue we can refactor it later into something else but I think it should be fine here.

Reply:
> I think it's useful to be able to listen to this event anywhere that's needed (like maybe the downloads manager can listen to this status changes to restart a download when the connection is back)

Gotcha. Makes sense. I'm thinking we could have a separate module that exports nodejs event emitters used for backend->backend communication. That way the frontend/preload script won't get excess functionality exposed to it and we don't have to worry about dependency issues since you'd just be importing the event emitter instead of the whole module + the event emitter.

What do you think?